### ISSUE TITLE
Long-term caching support

### ISSUE THREAD

**Issue by thangngoc89** [MoOx/phenomic#130]
Example : https://github.com/thangngoc89/statinamic-theme-lumen/commit/d76c0b851b2cf55f48d25ebd456ef6c70f12b88f

**Comment by MoOx** [comment]
I think we should use chunkhash, not hash.

**Comment by thangngoc89** [comment]
Yes.  That would help but not enough

**Comment by MoOx** [comment]
Why? [chunkhash] is supposed to not change if the content is the same.

**Comment by thangngoc89** [comment]
Even though code content is the same, webpack will inject its manifest (module id,  ... ) into chunk randomly. We need to extract those to a seperate file,  and inject them back to html.

Also,  babel-plugin-webpack...  injects classnames into chunk (which contains random hash)

**Comment by MoOx** [comment]
Oh yeah... True. So this won't be that easy :)

**Comment by thangngoc89** [comment]
We can have put common package such as react, redux, react-router into a different chunk to minize number of bytes to download each build

**Comment by thangngoc89** [comment]
alright, i succeed to split js into common chunk and main chunk. Common chunk won't change when its hash when changing markdown content. Main chunk' hash will keep changing because css hash. I think we can safely disable [hash] in css.

`[path][name]--[local]` is unique enogh.

Any thoughts ? @MoOx

**Comment by MoOx** [comment]
[hash] is the point of CSS module. Maybe there is a way to have a sort of [chunkhash] for css module too?

**Comment by thangngoc89** [comment]
I don't know,  you're an expert in postcss. By the way,  css-loader@^0.23.x (latest) broke dev server. It generate different hash in client and server

**Comment by MoOx** [comment]
The fuck. Is there an opened issue about that ?

**Comment by thangngoc89** [comment]
I don't know,  just tried to update everything to the latest version and got that

**Comment by thangngoc89** [comment]
Here is what to do for anyone who is interesting:

```js
// webpack.config.client.js

// ↓ HANDLE WITH CARE ↓ \\
  plugins: [
    ...webpackConfig.plugins,
    new webpack.optimize.CommonsChunkPlugin({
      names: ['statinamic-common', 'manifest'],
    }),
    // new webpack.NamedModulesPlugin(),
  ],
  output: {
    ...webpackConfig.output,
    libraryTarget: "var",
    filename: "[name].[hash].js",
    ...config.production && {
      filename: "[name].[chunkhash].js",
      chunkFilename: "[chunkhash].js",
    },
  },
  entry: {
    "statinamic-main": path.join(__dirname, "index-client"),
    "statinamic-common": [
      "react",
      "redux",
      "react-redux",
    ],
  },
```

Hashes won't change if markdown content changed. 

If app code changed, manifest and vendor code's hashes will change.

Result
```
4.0K manifest.a79b56b716403cda118c.js
144K statinamic-common.e5f7372fe5b4c362b0c4.js
128K statinamic-main.5088db94c01d82c1f41a.js
```

manifest chunk is small so it can be inline to prevent additional request. You can add more package to common chunk to have a smaller main chunk

**Comment by thangngoc89** [comment]
Since, there is no cache in github pages, this is useless for docs

**Comment by MoOx** [comment]
Not so sure, not everyone use gh pages. I plan to use sftp myself on an apache instance to get htaccess and real redirection.

**Comment by MoOx** [comment]
Btw, you can use cloudflare after github pages.

**Comment by thangngoc89** [comment]
I meant statinamic's docs site. Why when through all the trouble ? Cloudfront + s3 is a good and cheap service. Cloudflare is good too. https for free

**Comment by MoOx** [comment]
Oh yeah. But we could document that on the docs :)

**Comment by thangngoc89** [comment]
I think we could use [chunkhash] by default for boilerplate.

**Comment by thangngoc89** [comment]
1 problem : 

3 files above must be in this order: manifest --> common --> main

Need some refactor here 

https://github.com/MoOx/statinamic/blob/master/src%2Fstatic%2Fto-html%2Furl-as-html.js#L109-L114

**Comment by MoOx** [comment]
Name them 1-manifest, 2-... :D ?

**Comment by thangngoc89** [comment]
Haha. Work for me :+1:

**Comment by thangngoc89** [comment]
This is doable now tih 0.8.0. Just need some docs

**Comment by thangngoc89** [comment]
After all, to persist hash on vendor chunk. Your dependencies has to unchanged. Only work on local or CI with cached dependencies

**Comment by thangngoc89** [comment]
Updated an example in to first post :arrow_up:

**Comment by thangngoc89** [comment]
Closed by #411
